### ISSUE TITLE
Benchmarking Mobilenet inference performance fails on AMD GPU, Windows

### ISSUE THREAD

**Issue by swilcoxtx** [plaidml/plaidml#380]
Plaidml fails for plaidbench keras mobilenet test and plaidbench --batch-size 1 keras --train mobilenet test for AMD GPU (opencl_amd_oland). I thought it might be that AMD was not functioning at full capacity, so I increased the GPU_MAX_HEAP_SIZE and CPU_MAX_ALLOC_PERCENT to 95. I also tried using the Intel GPU (opencl_intel_hd_graphics). However, when running plaidml-setup.py I was never prompted to save settings to the .plaidml file. Is there something I need to update or install?

******* plaidml-setup.py OUTPUT *******
Default Config Devices:
   No devices.

Experimental Config Devices:
   opencl_amd_oland.0 : Advanced Micro Devices, Inc. Oland (OpenCL)
   llvm_cpu.0 : CPU (LLVM)
   opencl_intel_hd_graphics_4600.0 : Intel(R) Corporation Intel(R) HD Graphics 4600 (OpenCL)
   opencl_cpu.0 : Intel(R) Corporation CPU (OpenCL)

Using experimental devices can cause poor performance, crashes, and other nastiness.

Enable experimental device support? (y,n)[n]:y

Multiple devices detected (You can override by setting PLAIDML_DEVICE_IDS).
Please choose a default device:

   1 : opencl_amd_oland.0
   2 : llvm_cpu.0
   3 : opencl_intel_hd_graphics_4600.0
   4 : opencl_cpu.0

Default device? (1,2,3,4)[1]:1

Selected device:
    opencl_amd_oland.0

Almost done. Multiplying some matrices...
Tile code:
  function (B[X,Z], C[Z,Y]) -> (A) { A[x,y : X,Y] = +(B[x,z] * C[z,y]); }
Whew. That worked.

Save settings to C:\Users\*\.plaidml? (y,n)[y]:y
Success!

******* plaidbench keras mobilenet OUTPUT  *********
Running 1024 examples with mobilenet, batch size 1, on backend plaid
INFO:plaidml:Opening device "opencl_amd_oland.0"
Compiling network... Warming up... Running...
Example finished, elapsed: 12.564s (compile), 29.923s (execution)

-----------------------------------------------------------------------------------------
Network Name         Inference Latency         Time / FPS
-----------------------------------------------------------------------------------------
mobilenet            29.22 ms                  18.78 ms / 53.25 fps
Correctness: FAIL, max_error: 80.5865478515625, max_abs_error: 0.48130080103874207, fail_ratio: 0.929

*******plaidbench --batch-size 1 keras --train mobilenet  OUTPUT **********
(mlgpu) C:\Users\*>plaidbench --batch-size 1 keras --train mobilenet
Running 1024 examples with mobilenet, batch size 1, on backend plaid
Loading CIFAR data
Downloading data from https://www.cs.toronto.edu/~kriz/cifar-10-python.tar.gz
170500096/170498071 [==============================] - 103s 1us/step
INFO:plaidml:Opening device "opencl_amd_oland.0"
Compiling network...Epoch 1/1
INFO:plaidml:Analyzing Ops: 556 of 2281 operations complete
INFO:plaidml:Analyzing Ops: 1174 of 2281 operations complete
INFO:plaidml:Analyzing Ops: 1544 of 2281 operations complete
1/1 [==============================] - 76s 76s/step - loss: 7.1508 - acc: 0.0000e+00
 Warming up...Epoch 1/1
32/32 [==============================] - 7s 225ms/step - loss: 5.2037 - acc: 0.1875
 Running...
Epoch 1/1
   1/1024 [..............................] - ETA: 4:47 - loss: 4.1405 - acc: 0.000   2/1024 [..............................] - ETA: 5:11 - loss: 4.6125 - acc: 0.000   3/1024 [..............................] - ETA: 4:41 - loss: 3.1108 - acc: 0.333   4/1024 [..............................] - ETA: 4:26 - loss: 3.6381 - a
......
982/1024 [===========================>..] - ETA: 10s - loss: 3.8837 - acc: 0.1222 983/1024 [===========================>..] - ETA: 9s - loss: 3.8817 - acc: 0.1221 1024/1024 [==============================] - 246s 241ms/step - loss: 3.8450 - acc: 0.1221
Example finished, elapsed: 76.294s (compile), 246.381s (execution)

-----------------------------------------------------------------------------------------
Network Name         Inference Latency         Time / FPS
-----------------------------------------------------------------------------------------
mobilenet            240.61 ms                 83.53 ms / 11.97 fps
Correctness: FAIL, max_error: 0.424050731558952, max_abs_error: 3.032308434485458, fail_ratio: 1.0

Output of clinfo: 
Number of platforms:                             2
  Platform Profile:                              FULL_PROFILE
  Platform Version:                              OpenCL 2.1 AMD-APP (2906.10)
  Platform Name:                                 AMD Accelerated Parallel Processing
  Platform Vendor:                               Advanced Micro Devices, Inc.
  Platform Extensions:                           cl_khr_icd cl_khr_d3d10_sharing cl_khr_d3d11_sharing cl_khr_dx9_media_sharing cl_amd_event_callback cl_amd_offline_devices
  Platform Profile:                              FULL_PROFILE
  Platform Version:                              OpenCL 1.2
  Platform Name:                                 Intel(R) OpenCL
  Platform Vendor:                               Intel(R) Corporation
  Platform Extensions:                           cl_intel_dx9_media_sharing cl_khr_3d_image_writes cl_khr_byte_addressable_store cl_khr_d3d11_sharing cl_khr_depth_images cl_khr_dx9_media_sharing cl_khr_gl_sharing cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_icd cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_spir

  Platform Name:                                 AMD Accelerated Parallel Processing
Number of devices:                               1
  Device Type:                                   CL_DEVICE_TYPE_GPU
  Vendor ID:                                     1002h
  Board name:                                    AMD Radeon HD 8790M
  Device Topology:                               PCI[ B#1, D#0, F#0 ]
  Max compute units:                             6
  Max work items dimensions:                     3
    Max work items[0]:                           1024
    Max work items[1]:                           1024
    Max work items[2]:                           1024
  Max work group size:                           256
  Preferred vector width char:                   4
  Preferred vector width short:                  2
  Preferred vector width int:                    1
  Preferred vector width long:                   1
  Preferred vector width float:                  1
  Preferred vector width double:                 1
  Native vector width char:                      4
  Native vector width short:                     2
  Native vector width int:                       1
  Native vector width long:                      1
  Native vector width float:                     1
  Native vector width double:                    1
  Max clock frequency:                           900Mhz
  Address bits:                                  32
  Max memory allocation:                         1785095782
  Image support:                                 Yes
  Max number of images read arguments:           128
  Max number of images write arguments:          8
  Max image 2D width:                            16384
  Max image 2D height:                           16384
  Max image 3D width:                            2048
  Max image 3D height:                           2048
  Max image 3D depth:                            2048
  Max samplers within kernel:                    16
  Max size of kernel argument:                   1024
  Alignment (bits) of base address:              2048
  Minimum alignment (bytes) for any datatype:    128
  Single precision floating point capability
    Denorms:                                     No
    Quiet NaNs:                                  Yes
    Round to nearest even:                       Yes
    Round to zero:                               Yes
    Round to +ve and infinity:                   Yes
    IEEE754-2008 fused multiply-add:             Yes
  Cache type:                                    Read/Write
  Cache line size:                               64
  Cache size:                                    16384
  Global memory size:                            2039480320
  Constant buffer size:                          65536
  Max number of constant args:                   8
  Local memory type:                             Scratchpad
  Local memory size:                             32768
  Max pipe arguments:                            0
  Max pipe active reservations:                  0
  Max pipe packet size:                          0
  Max global variable size:                      0
  Max global variable preferred total size:      0
  Max read/write image args:                     0
  Max on device events:                          0
  Queue on device max size:                      0
  Max on device queues:                          0
  Queue on device preferred size:                0
  SVM capabilities:
    Coarse grain buffer:                         No
    Fine grain buffer:                           No
    Fine grain system:                           No
    Atomics:                                     No
  Preferred platform atomic alignment:           0
  Preferred global atomic alignment:             0
  Preferred local atomic alignment:              0
  Kernel Preferred work group size multiple:     64
  Error correction support:                      0
  Unified memory for Host and Device:            0
  Profiling timer resolution:                    1
  Device endianess:                              Little
  Available:                                     Yes
  Compiler available:                            Yes
  Execution capabilities:
    Execute OpenCL kernels:                      Yes
    Execute native function:                     No
  Queue on Host properties:
    Out-of-Order:                                No
    Profiling :                                  Yes
  Queue on Device properties:
    Out-of-Order:                                No
    Profiling :                                  No
  Platform ID:                                   00007FFF38228FD0
  Name:                                          Oland
  Vendor:                                        Advanced Micro Devices, Inc.
  Device OpenCL C version:                       OpenCL C 1.2
  Driver version:                                2906.10
  Profile:                                       FULL_PROFILE
  Version:                                       OpenCL 1.2 AMD-APP (2906.10)
  Extensions:                                    cl_khr_fp64 cl_amd_fp64 cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_int64_base_atomics cl_khr_int64_extended_atomics cl_khr_3d_image_writes cl_khr_byte_addressable_store cl_khr_gl_sharing cl_amd_device_attribute_query cl_amd_vec3 cl_amd_printf cl_amd_media_ops cl_amd_media_ops2 cl_amd_popcnt cl_khr_d3d10_sharing cl_khr_d3d11_sharing cl_khr_dx9_media_sharing cl_khr_image2d_from_buffer cl_khr_spir cl_khr_gl_event cl_amd_liquid_flash

  Platform Name:                                 Intel(R) OpenCL
Number of devices:                               2
  Device Type:                                   CL_DEVICE_TYPE_GPU
  Vendor ID:                                     8086h
  Max compute units:                             20
  Max work items dimensions:                     3
    Max work items[0]:                           512
    Max work items[1]:                           512
    Max work items[2]:                           512
  Max work group size:                           512
  Preferred vector width char:                   1
  Preferred vector width short:                  1
  Preferred vector width int:                    1
  Preferred vector width long:                   1
  Preferred vector width float:                  1
  Preferred vector width double:                 0
  Native vector width char:                      1
  Native vector width short:                     1
  Native vector width int:                       1
  Native vector width long:                      1
  Native vector width float:                     1
  Native vector width double:                    0
  Max clock frequency:                           1300Mhz
  Address bits:                                  64
  Max memory allocation:                         427189862
  Image support:                                 Yes
  Max number of images read arguments:           128
  Max number of images write arguments:          128
  Max image 2D width:                            16384
  Max image 2D height:                           16384
  Max image 3D width:                            2048
  Max image 3D height:                           2048
  Max image 3D depth:                            2048
  Max samplers within kernel:                    16
  Max size of kernel argument:                   1024
  Alignment (bits) of base address:              1024
  Minimum alignment (bytes) for any datatype:    128
  Single precision floating point capability
    Denorms:                                     No
    Quiet NaNs:                                  Yes
    Round to nearest even:                       Yes
    Round to zero:                               Yes
    Round to +ve and infinity:                   Yes
    IEEE754-2008 fused multiply-add:             No
  Cache type:                                    Read/Write
  Cache line size:                               64
  Cache size:                                    262144
  Global memory size:                            1708759450
  Constant buffer size:                          65536
  Max number of constant args:                   8
  Local memory type:                             Scratchpad
  Local memory size:                             65536
  Kernel Preferred work group size multiple:     32
  Error correction support:                      0
  Unified memory for Host and Device:            1
  Profiling timer resolution:                    80
  Device endianess:                              Little
  Available:                                     Yes
  Compiler available:                            Yes
  Execution capabilities:
    Execute OpenCL kernels:                      Yes
    Execute native function:                     No
  Queue on Host properties:
    Out-of-Order:                                No
    Profiling :                                  Yes
  Platform ID:                                   000001DD599534D0
  Name:                                          Intel(R) HD Graphics 4600
  Vendor:                                        Intel(R) Corporation
  Device OpenCL C version:                       OpenCL C 1.2
  Driver version:                                20.19.15.4835
  Profile:                                       FULL_PROFILE
  Version:                                       OpenCL 1.2
  Extensions:                                    cl_intel_accelerator cl_intel_advanced_motion_estimation cl_intel_ctz cl_intel_d3d11_nv12_media_sharing cl_intel_dx9_media_sharing cl_intel_motion_estimation cl_intel_simultaneous_sharing cl_intel_subgroups cl_khr_3d_image_writes cl_khr_byte_addressable_store cl_khr_d3d10_sharing cl_khr_d3d11_sharing cl_khr_depth_images cl_khr_dx9_media_sharing cl_khr_gl_depth_images cl_khr_gl_event cl_khr_gl_msaa_sharing cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_gl_sharing cl_khr_icd cl_khr_image2d_from_buffer cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_spir

  Device Type:                                   CL_DEVICE_TYPE_CPU
  Vendor ID:                                     8086h
  Max compute units:                             8
  Max work items dimensions:                     3
    Max work items[0]:                           8192
    Max work items[1]:                           8192
    Max work items[2]:                           8192
  Max work group size:                           8192
  Preferred vector width char:                   1
  Preferred vector width short:                  1
  Preferred vector width int:                    1
  Preferred vector width long:                   1
  Preferred vector width float:                  1
  Preferred vector width double:                 1
  Native vector width char:                      32
  Native vector width short:                     16
  Native vector width int:                       8
  Native vector width long:                      4
  Native vector width float:                     8
  Native vector width double:                    4
  Max clock frequency:                           2700Mhz
  Address bits:                                  64
  Max memory allocation:                         4270131200
  Image support:                                 Yes
  Max number of images read arguments:           480
  Max number of images write arguments:          480
  Max image 2D width:                            16384
  Max image 2D height:                           16384
  Max image 3D width:                            2048
  Max image 3D height:                           2048
  Max image 3D depth:                            2048
  Max samplers within kernel:                    480
  Max size of kernel argument:                   3840
  Alignment (bits) of base address:              1024
  Minimum alignment (bytes) for any datatype:    128
  Single precision floating point capability
    Denorms:                                     Yes
    Quiet NaNs:                                  Yes
    Round to nearest even:                       Yes
    Round to zero:                               No
    Round to +ve and infinity:                   No
    IEEE754-2008 fused multiply-add:             No
  Cache type:                                    Read/Write
  Cache line size:                               64
  Cache size:                                    262144
  Global memory size:                            17080524800
  Constant buffer size:                          131072
  Max number of constant args:                   480
  Local memory type:                             Global
  Local memory size:                             32768
  Kernel Preferred work group size multiple:     128
  Error correction support:                      0
  Unified memory for Host and Device:            1
  Profiling timer resolution:                    100
  Device endianess:                              Little
  Available:                                     Yes
  Compiler available:                            Yes
  Execution capabilities:
    Execute OpenCL kernels:                      Yes
    Execute native function:                     Yes
  Queue on Host properties:
    Out-of-Order:                                Yes
    Profiling :                                  Yes
  Platform ID:                                   000001DD599534D0
  Name:                                          Intel(R) Core(TM) i7-4800MQ CPU @ 2.70GHz
  Vendor:                                        Intel(R) Corporation
  Device OpenCL C version:                       OpenCL C 1.2
  Driver version:                                5.2.0.10094
  Profile:                                       FULL_PROFILE
  Version:                                       OpenCL 1.2 (Build 10094)
  Extensions:                                    cl_khr_icd cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_byte_addressable_store cl_khr_depth_images cl_khr_3d_image_writes cl_intel_exec_by_local_thread cl_khr_spir cl_khr_dx9_media_sharing cl_intel_dx9_media_sharing cl_khr_d3d11_sharing cl_khr_gl_sharing cl_khr_fp64

AMD Settings:  Radeon Software Version 19.7.4
Windows 10

**Comment by mwyi** [comment]
Reproduced on our end. Will continue to debug.

**Comment by zhanwenchen** [comment]
Similar issue after updating to 0.6.4 on Mac.

**Comment by themrwhitehair** [comment]
was this ever resolved?

Correctness: FAIL, max_error: 80.5865478515625, max_abs_error: 0.48130080103874207, fail_ratio: 0.929

i have the same issue on my amd card

**Comment by swilcoxtx** [comment]
I was never able to fix it and never heard back. Still unresolved...

**Comment by themrwhitehair** [comment]
> I was never able to fix it and never heard back. Still unresolved...

I was able to find a work around. i ran my program in a setting that used less memory and it was able to work. what kind of graphics card do you have? i might suggest upgrading.

**Comment by HOZHENWAI** [comment]
Similar issue, inference is ok but training fails. Is there a way to troubleshoot this issue? How about checking for older version of PlaidML?
